### PR TITLE
[flutter_markdown] Fix WidgetSpan Support in MarkdownElementBuilder

### PR DESCRIPTION
**Pull request by pzierahn**
This pull request addresses a critical issue identified in the `flutter_markdown` package, specifically relating to the handling of `WidgetSpan` elements within `Text.rich` elements when utilized inside `MarkdownElementBuilder`s. Prior to this fix, the inclusion of `WidgetSpan` instances within markdown content led to casting issues, disrupting the normal rendering flow of the markdown content.

Key Changes:
- Resolved casting problems associated with using `WidgetSpan` inside `Text.rich` elements by refining the type handling and span merging logic within the affected functions.
- Thoroughly commented and cleaned up the code within the error-prone functions to enhance readability and maintainability, ensuring that future modifications can be made more easily.
- Adjusted existing tests to align with the newly introduced span merging logic, which is now more efficient and produces more predictable outcomes.
- Introduced new tests specifically designed to cover scenarios where `WidgetSpan` elements are included within markdown content, ensuring that this issue does not resurface in future updates.

The adjustments made in this pull request ensure that developers utilizing the `flutter_markdown` package can now seamlessly incorporate `WidgetSpan` elements within their markdown content without encountering the previously observed casting issues. This fix not only improves the package's robustness but also expands its flexibility, allowing for richer text compositions within markdown.

This pull request closes issue [#144383](https://github.com/flutter/flutter/issues/144383), effectively addressing the reported bug and enhancing the overall functionality of the `flutter_markdown` package.

I'm welcoming feedback and suggestions for improvement.

Fixed issues:
* [#144383](https://github.com/flutter/flutter/issues/144383)

### CODE REVIEW

**Comment by domesticmouse** [comment]
PTAL @stuartmorgan

File: packages/flutter_markdown/lib/src/builder.dart [hunk 1]
```diff
@@ -707,52 +707,111 @@ class MarkdownBuilder implements md.NodeVisitor {
     }
   }
 
+  /// Extracts all spans from an inline element and merges them into a single list
+  List<InlineSpan> _getInlineSpans(InlineSpan span) {
+    final List<InlineSpan> spans = <InlineSpan>[];
+
+    if (span is TextSpan && span.children != null) {
```
[the remark this answers is not in the document]
Comment (pzierahn) [reply packages/flutter_markdown/lib/src/builder.dart:714]: It appears that the previous implementation did not fully account for inheriting the parent span's styling onto its children. To ensure child spans inherit the parent's style, I can add the following code snippet:

```dart
// Merge the style of the parent with the style of the children
final List<InlineSpan> children = span.children!.map((InlineSpan childSpan) {
  if (childSpan is TextSpan) {
    return TextSpan(
      text: childSpan.text,
      recognizer: childSpan.recognizer,
      semanticsLabel: childSpan.semanticsLabel,
      style: childSpan.style?.merge(span.style),
    );
  } else {
    return childSpan;
  }
}).toList();

spans.addAll(children);
```

This code ensures the children `TextSpan`s properly inherit and merge the styling from the parent `TextSpan`, preserving the intended styling hierarchy.

**Comment by pzierahn** [comment]
@stuartmorgan All checks pass now

File: packages/flutter_markdown/lib/src/builder.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (domesticmouse) [reply packages/flutter_markdown/lib/src/builder.dart:714]: sgtm!

**Comment by pzierahn** [comment]
@domesticmouse I have implemented the parent child style matching.

File: packages/flutter_markdown/test/inline_widget_test.dart [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+class SubscriptBuilder extends MarkdownElementBuilder {
+  @override
+  Widget visitElementAfterWithContext(
+    BuildContext context,
+    md.Element element,
+    TextStyle? preferredStyle,
+    TextStyle? parentStyle,
+  ) {
+    // This pull request will add support for WidgetSpan in the FlutterMarkdown package
```
Comment (tarrinneal) [inline packages/flutter_markdown/test/inline_widget_test.dart:61]: This comment should be removed.

**Review by tarrinneal — CHANGES_REQUESTED** [review]
one quick change

**Review by tarrinneal — APPROVED** [review]
thanks!

**Pull request merged**